File: app/src/main/java/org/movieos/feeder/utilities/RealmAdapter.java
package org.movieos.feeder.utilities;

import android.databinding.ViewDataBinding;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import org.movieos.feeder.model.IntegerPrimaryKey;

import java.lang.reflect.InvocationTargetException;
import java.util.List;

import io.realm.OrderedCollectionChangeSet;
import io.realm.OrderedRealmCollectionChangeListener;
import io.realm.RealmObject;
import io.realm.RealmResults;


public abstract class RealmAdapter<T extends RealmObject & IntegerPrimaryKey, B extends ViewDataBinding> extends RecyclerView.Adapter<RealmAdapter.FeedViewHolder<B>> implements OrderedRealmCollectionChangeListener<RealmResults<T>> {

    Class<B> mKlass;
    RealmResults<T> mQuery;

    public RealmAdapter(Class<B> klass, RealmResults<T> query) {
        mKlass = klass;
        mQuery = query;
        setHasStableIds(true);
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        mQuery.addChangeListener(this);
        if (recyclerView.getLayoutManager() == null) {
            recyclerView.setLayoutManager(new LinearLayoutManager(recyclerView.getContext()));
        }
    }

    public void setQuery(RealmResults<T> query) {
        if (mQuery != null) {
            mQuery.removeChangeListener(this);
            mQuery.removeChangeListener(this);
        }
        mQuery = query;
        mQuery.addChangeListener(this);
        notifyDataSetChanged();
    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        mQuery.removeChangeListener(this);
    }

    @Override
    public void onChange(RealmResults<T> collection, OrderedCollectionChangeSet changeSet) {
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return mQuery.size();
    }

    @Override
    public long getItemId(int position) {
        return mQuery.get(position).getId();
    }

    public List<Integer> getIds() {
        //noinspection Convert2MethodRef
        return ListUtils.map(t -> t.getId(), mQuery);
    }

    @SuppressWarnings("unchecked")
    @Override
    public FeedViewHolder<B> onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        try {
            B binding = (B) mKlass.getMethod("inflate", LayoutInflater.class, ViewGroup.class, boolean.class).invoke(null, inflater, parent, false);
            return new FeedViewHolder<>(binding);
        } catch (IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
            throw new RuntimeException("Failed to instantiate " + mKlass, e);
        }
    }

    @Override
    public void onBindViewHolder(FeedViewHolder<B> holder, int position) {
        T instance = mQuery.get(position);
        onBindViewHolder(holder, instance);
    }

    public abstract void onBindViewHolder(FeedViewHolder<B> holder, T instance);

    public final static class FeedViewHolder<B extends ViewDataBinding> extends RecyclerView.ViewHolder {
        @NonNull
        private B mBinding;

        protected FeedViewHolder(@NonNull B binding) {
            super(binding.getRoot());
            mBinding = binding;
        }

        @NonNull
        public B getBinding() {
            return mBinding;
        }
    }
}

File: app/src/main/java/org/movieos/feeder/utilities/RealmAdapter.kt
package org.movieos.feeder.utilities

import android.databinding.ViewDataBinding
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import io.realm.OrderedCollectionChangeSet
import io.realm.OrderedRealmCollectionChangeListener
import io.realm.RealmObject
import io.realm.RealmResults
import org.movieos.feeder.model.IntegerPrimaryKey
import java.lang.reflect.InvocationTargetException


abstract class RealmAdapter<T, B>(
        internal var mKlass: Class<B>,
        internal var mQuery: RealmResults<T>
) : RecyclerView.Adapter<RealmAdapter.FeedViewHolder<B>>(),
        OrderedRealmCollectionChangeListener<RealmResults<T>>
where T : RealmObject, T : IntegerPrimaryKey, B : ViewDataBinding {

    init {
        setHasStableIds(true)
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        mQuery.addChangeListener(this)
        if (recyclerView.layoutManager == null) {
            recyclerView.layoutManager = LinearLayoutManager(recyclerView.context)
        }
    }

    fun setQuery(query: RealmResults<T>) {
        mQuery.removeChangeListener(this)
        query.addChangeListener(this)
        mQuery = query
        notifyDataSetChanged()
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView?) {
        super.onDetachedFromRecyclerView(recyclerView)
        mQuery.removeChangeListener(this)
    }

    override fun onChange(collection: RealmResults<T>, changeSet: OrderedCollectionChangeSet) {
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return mQuery.size
    }

    override fun getItemId(position: Int): Long {
        return mQuery[position].id.toLong()
    }

    val ids: List<Int>
        get() = mQuery.map { t -> t.id }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FeedViewHolder<B> {
        val inflater = LayoutInflater.from(parent.context)
        try {
            val binding = mKlass.getMethod("inflate", LayoutInflater::class.java, ViewGroup::class.java, Boolean::class.javaPrimitiveType).invoke(null, inflater, parent, false) as B
            return FeedViewHolder(binding)
        } catch (e: IllegalAccessException) {
            throw RuntimeException("Failed to instantiate " + mKlass, e)
        } catch (e: NoSuchMethodException) {
            throw RuntimeException("Failed to instantiate " + mKlass, e)
        } catch (e: InvocationTargetException) {
            throw RuntimeException("Failed to instantiate " + mKlass, e)
        }

    }

    override fun onBindViewHolder(holder: FeedViewHolder<B>, position: Int) {
        val instance = mQuery[position]
        onBindViewHolder(holder, instance)
    }

    abstract fun onBindViewHolder(holder: FeedViewHolder<B>, instance: T)

    class FeedViewHolder<B : ViewDataBinding>(val binding: B) : RecyclerView.ViewHolder(binding.root)
}
